File: algorithms/[20]有效的括号/solution.ts
// topics = ["字符串", "栈"]

function isValid(s: string): boolean {
  const stack: String[] = [];

  for (let c of s) {
    const last = stack[stack.length - 1];
    if ((c == ')' && last == '(') || (c == ']' && last == '[') || (c == '}' && last == '{')) {
      stack.pop();
    } else {
      stack.push(c);
    }
  }

  return stack.length == 0;
}

File: algorithms/[20]有效的括号/Solution.cs
// topics = ['字符串", "栈"]

using System.Collections.Generic;

public class Solution
{
    public bool IsValid(string str)
    {
        Stack<char> s = new Stack<char>();
        Dictionary<char, char> d = new Dictionary<char, char>()
        {
            { ')', '(' },
            { ']', '[' },
            { '}', '{' },
        };

        foreach (char ch in str)
        {
            if (d.ContainsKey(ch))
            {
                if (s.Count == 0 || s.Peek() != d[ch])
                {
                    return false;
                }
                s.Pop();
            }
            else
            {
                s.Push(ch);
            }
        }

        return s.Count == 0;
    }
}
